Align quality-on-demand with Commonalties regarding optional device object and error messages

hdamker opened the pull request:
#### What type of PR is this?

Add one of the following kinds:
* correction
* documentation

#### What this PR does / why we need it (Updated):

* Add documentation about "Handling of device information" within the API description. Made device parameter optional within createSession
* Aligned Device object and info object with Commonalities
* Updated errorMessages according to Commonalities.
  * Added 422, removed 501 (as createSession is not optional within the API, and then 501 is not any more used)
  * Replaced 400 error response examples with Generic400 from Commonalities, introduced specific error code QUALITY_ON_DEMAND.DURATION_OUT_OF_RANGE

#### Which issue(s) this PR fixes:

<!-- Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`. -->

Fixes #313
Fixes #315 
Fixes #316 

#### Special notes for reviewers:

~~Intentionally only done for quality-on-demand to reduce potential merge conflicts.~~
Done for both APIs quality-on-demand and qos-profiles.

#### Changelog input (Updated)

```
Made device object optional within quality-on-demand, added documentation according to Commonalities
Updated error messages according to Commonalities, especially error messages related to device object
Introduced specific error code QUALITY_ON_DEMAND.DURATION_OUT_OF_RANGE

```

hdamker commented:
@maxl2287 @akoshunyadi @jlurien I created the PR by intention first as draft, but would be great if you can already review. 
I would like to rebase if #325 is merged (which is also correcting some errorMessages). 

@eric-murray the documentation I copied from Commonalities is only referring to "device" associated with the access token. I get your point that "end user" is more correct. But do you have a proposal not to confuse API consumers here? My proposal to put "(and device)" behind end user is most probably not the best ...

jlurien commented on code/API_definitions/quality-on-demand.yaml:
```diff
@@ -871,46 +881,73 @@ components:
+                code: INVALID_TOKEN_CONTEXT
+                message: "{{field}} is not consistent with access token."
 
     SessionNotFound404:
```
Some other cases for 404 from Commonalities must also be included

hdamker replied:
I have by intention not touched "SessionNotFound404" as there is now in addition "Generic404" in #325.

Would you add the cases from Commonalities in the definition?

jlurien commented on code/API_definitions/quality-on-demand.yaml:
```diff
@@ -940,6 +977,29 @@ components:
             code: CONFLICT
             message: Conflict with an existing session for the same device.
 
+    Generic422:
```
We probably have to add the new UNIDENTIFIABLE_DEVICE (still in PR in Commonalities)

hdamker replied:
Right ... I just copied from artifacts/CAMARA_common.yaml here.

hdamker replied:
Copied now the addition from https://github.com/camaraproject/Commonalities/pull/256/files

jlurien commented:
@hdamker are you planning to remove the draft status? This PR is key for the meta-release as it incorporates the alignments to guidelines. Once this is merged, we can adapt the remaining ones (listSessions, Provisioning) to the updated content. For listSessions I have to define same behaviour for device optionality and adopt the new errors.

hdamker commented:
@jlurien Yes, removed the draft status now to be able to discuss it within the meeting. We should agree who (which PR) is changing the 404 errors.

hdamker commented:
As discussed within QoD Call:

- [ ] Incorporate the Generic404 definition
- [ ] Add UNIDENTIFIABLE_DEVICE for 422
- [ ] Differentiate SessionInfo for request and response (for #328) and add proposal from #328
- [ ] Apply respective changes to dos-sessions

hdamker commented:
@camaraproject/quality-on-demand_codeowners @camaraproject/quality-on-demand_maintainers 

I hope that I addressed all points which we have discussed. Please review - it would be good to the PR merged on Wednesday, July 31st.

jlurien commented on code/API_definitions/quality-on-demand.yaml:
```diff
@@ -197,24 +225,6 @@ paths:
                     status: 400
                     code: INVALID_ARGUMENT
                     message: "Expected property is missing: device"
-                InsufficientDeviceProperties:
```
We should deprecate as well DeviceMissing, which is covered with better dedicated errors

hdamker replied:
Agreed, done as part of https://github.com/camaraproject/QualityOnDemand/pull/326/commits/061776de7e60d68b91967e815413adc88e37651b

jlurien commented on code/API_definitions/quality-on-demand.yaml:
```diff
@@ -197,24 +225,6 @@ paths:
-                    status: 400
-                    code: INVALID_ARGUMENT
-                    message: "Unable to identify device from specified parameters: device"
                 InvalidDevicePublicPortValue:
```
In general I think we may rely on the generic 400 for errors when input does not comply with schema: mandatory property missing, property values out of range, etc

jlurien replied:
Yes, the DurationOutOfRangeForQoSProfile case is relevant and not covered by the compliance to schema. In these cases  we can design an API dedicated code, e.g.  QUALITY_ON_DEMAND.DURATION_OUT_OF_RANGE_FOR_QOS_PROFILE. Or something shorter, but API_NAME.ERROR_CODE is the expected format.

hdamker replied:
Will do so. Just a question if understood it correctly:

Looking on CAMARA_common, the template is:

```
            GENERIC_400_{{SPECIFIC_CODE}}:
              description: Specific Syntax Exception regarding a field that is relevant in the context of the API
              value:
                status: 400
                code: { { SPECIFIC_CODE } }
                message: { { SPECIFIC_CODE_MESSAGE } }
```

merging this with 

```
            DurationOutOfRangeForQoSProfile:
              summary: The requested duration is out of the allowed range for the specific QoS profile
              value:
                status: 400
                code: OUT_OF_RANGE
                message: "The requested duration is out of the allowed range for the specific QoS profile"
```

my proposal would be ("GENERIC_400_{{SPECIFIC_CODE}}" does not make sense to me, and I haven't found it anywhere):

```
            DurationOutOfRangeForQoSProfile:
              description: The requested duration is out of the allowed range for the specific QoS profile
              value:
                status: 400
                code: QUALITY_ON_DEMAND.DURATION_OUT_OF_RANGE
                message: The requested duration is out of the allowed range for the specific QoS profile
```

ok?

hdamker replied:
QUALITY_ON_DEMAND.DURATION_OUT_OF_RANGE introduced in https://github.com/camaraproject/QualityOnDemand/pull/326/commits/5c9bddea1a3487d98bd59f89c7492ea45bf0c4c6

jlurien commented on code/API_definitions/quality-on-demand.yaml:
```diff
@@ -508,12 +524,15 @@ components:
-      description: Session related information.
+      description: |
+        Session related information returned in responses.
+        Within the optional device object only one identifier will be returned (with the original value provided in CreateSession).
```
```suggestion
        Device object only to be returned if provided in createSession, and with the same identifier values that were provided originally.
```
We may provide several identifiers in device in the request, so it would be confusing saying that only one must be returned in that case

hdamker replied:
@jlurien I followed here the proposal from @eric-murray in https://github.com/camaraproject/QualityOnDemand/issues/328#issuecomment-2253079545:

> in fact, I would go further, and say that only one "identifier" (it might be a value pair) should be in the session info, which identifier being decided by the API provider. This is because of https://github.com/camaraproject/Commonalities/issues/259. We cannot now return 422 - DEVICE_IDENTIFIERS_MISMATCH (otherwise the API consumer will learn that a given ipv4address is not being used by a given phoneNumber), so better to accept one and ignore the rest.

So the one returned identifier would be the one used to identify the device, without giving further information about the other identifiers. If we would return unsupported identifiers (e.g. IPv4 address/port value pair) it would be also confusing.

jlurien replied:
Even if we go for Eric's proposal, the current text is ambiguous to me. Is the optional device to be returned if present in createSession request, or is it up to the implementation to return it?

Assuming that implementation considers only one and ignores the rest, I would reinforce that, something like:
```
Optional device object only to be returned if provided in createSession. If more that one type of device identifier was provided, only one identifier will be returned (at implementation choice and with the original value provided in createSession)

```

hdamker replied:
@jlurien I like your proposal and have applied it in [xxx](https://github.com/camaraproject/QualityOnDemand/pull/326/commits/27a767294c8ae8b959bf67f26a795ea09c7fbd30). I deleted also the now redundant sentence from line 464.

jlurien commented on code/API_definitions/quality-on-demand.yaml:
```diff
@@ -508,12 +524,15 @@ components:
+        Session related information returned in responses.
+        Within the optional device object only one identifier will be returned (with the original value provided in CreateSession).
+        If no device identifier was provided in CreateSession (together with a 3-legged token), no device object will be returned in SessionInfo.
+        Please note that IP addresses of devices can change and get reused and can't be used in these cases again to identify the same device.
```
```suggestion
        Please note that IP addresses of devices can change and get reused, so the original values may no longer identify the same device. They identified the device at the time of session creation.
```

jlurien reviewed (commented):
For qos-profiles, device errors look now strange but they should make sense once we merge #318 

Some comments but overall looks good. Thanks

jlurien commented:
@hdamker I'll start applying the same changes to retrieveSessions and qod-provisioning PRs, even if this one is still not merged. I will double check with the final version.

hdamker commented on code/API_definitions/qos-profiles.yaml:
```diff
@@ -399,46 +427,73 @@ components:
+                code: INVALID_TOKEN_CONTEXT
+                message: "{{field}} is not consistent with access token."
 
     QosProfilesNotFound404:
```
Done in https://github.com/camaraproject/QualityOnDemand/pull/326/commits/033602e7dab34a2fbdac874e1a4524e84a5a8909

hdamker commented:
@jlurien @RandyLevensalor Please review for consistency with your PRs.

Beyond that I suppose that we should merge this PR asap to get then #318 and #325 done.

@eric-murray please review as well and have a look on the open conversation https://github.com/camaraproject/QualityOnDemand/pull/326#discussion_r1696774175. Happy to use a better wording which reflects your proposal.

jlurien commented on code/API_definitions/quality-on-demand.yaml:
```diff
@@ -269,7 +229,9 @@ paths:
-        - The access token may be either 2-legged or 3-legged. If a 3-legged access token is used, the end user associated with the session must also be associated with the access token.
+        - The access token may be either 2-legged or 3-legged.
+        - If a 3-legged access token is used, the end user (and device) associated with the session must also be associated with the access token.
+        - The session must must have been created by the same API client given in the access token
```
```suggestion
        - The session must have been created by the same API client given in the access token
```

jlurien commented on code/API_definitions/quality-on-demand.yaml:
```diff
@@ -108,7 +136,7 @@ paths:
 
           This behavior allows clients which are not receiving notification events but are polling to get the session information with the `qosStatus` `UNAVAILABLE` and `statusInfo` `NETWORK_TERMINATED`. Before a client can attempt to create a new QoD session for the same device and flow period they must release the session resources with an explicit `delete` operation if not yet automatically de ...
-        - The access token may be either 2-legged or 3-legged. If a 3-legged access token is used, the end user identified by the `device` parameter must also be associated with the access token.
+        - The access token may be either 2-legged or 3-legged. If a 3-legged access token which is associated with a device is used, it is recommended NOT to include the `device` parameter in the request (see "Handling of device information" within the API description for details).
```
We may also add, as in retrieveSessions:

- If a 2-legged access token is used, the device parameter must be provided and identify a device.

hdamker replied:
Added in https://github.com/camaraproject/QualityOnDemand/pull/326/commits/8d23c566514d2f87e82e4a1741046e7b329c2d63

hdamker commented:
@jlurien Thanks for your latest comments, I've addressed them

@camaraproject/quality-on-demand_codeowners The PR should be ready for final review. Tried also to update the PR description with the actual changes.

jlurien reviewed (dismissed):
I found some other things to fix while adjusting and comparing with QoD Provisioning

jlurien commented:
Sorry, another part to fix:

externalDocs:
  description: Product documentation at Camara
  url: https://github.com/camaraproject/

should be

externalDocs:
  description: Project documentation at CAMARA
  url: https://github.com/camaraproject/QualityOnDemand

hdamker commented:
> externalDocs:
> description: Project documentation at CAMARA
> url: https://github.com/camaraproject/QualityOnDemand

ExternalDocs object updated in both APIs (I had planned it for the release PR).

> * createSession should have 404 error

Ah, yes - due to the new `DEVICE_NOT_FOUND` - Thanks!

maxl2287 commented on code/API_definitions/qos-profiles.yaml:
```diff
@@ -24,17 +24,45 @@ info:
+
+    ### Optional device object for 3-legged tokens:
+
+    - When using a 3-legged access token, the device associated with the access token must be considered as the device for the API request. This means that the device object is not required in the request, and if included it must identify the same device, therefore **it is recommended NOT to include it in these scenarios** to simplify the API usage and avoid additional validations.
```
Is this optional "device" neccessarry for qos-profiles?
As there is no need of having device-identifiers for retrieving qos-profile-information

>device associated with the access token must be considered as the device for the API request.

As there is no device in any qos-profile request, this sentence can at least removed.


wdyt?

maxl2287 replied:
Or are we just preparing for the future that there could be an option of processing device-data?

hdamker replied:
It's preparation for the very near future, more specifically for #318

jlurien commented:
Another ones:

- org.camaraproject.qod.v0.qos-status-changed -> org.camaraproject.quality-on-demand.vwip.qos-status-changed

Long but it should be the new API name, if we finally keep the proposal. 
vwip should be changed to v0 for the rc PR but as we know it will be v0 it doesn't have an impact

- DurationOutOfRangeForQoSProfile should not be part of Generic400 as that one should only contain the common ones. A separare 400 response would be needed for the operations that receive a duration

- In #299 I have considered a Generic404 with just NOT_FOUND and GenericDevice404 with both NOT_FOUND and DEVICE_NOT_FOUND, and I have applied the latest only to operations that may expect a device (and also 422 applies to them)

hdamker commented:
> * org.camaraproject.qod.v0.qos-status-changed -> org.camaraproject.quality-on-demand.vwip.qos-status-changed

That outside of the scope of this PR, at least we should open an issue for that to make the change more transparent.
Is there a guideline regarding the versioning of events which is different from the API versioning? Otherwise it would be in the release v0.11rc1, but I'm not sure if this is correct.

hdamker commented:
> * DurationOutOfRangeForQoSProfile should not be part of Generic400 as that one should only contain the common ones. A separare 400 response would be needed for the operations that receive a duration
> * In [New API QOD Provisioning #299](https://github.com/camaraproject/QualityOnDemand/pull/299) I have considered a Generic404 with just NOT_FOUND and GenericDevice404 with both NOT_FOUND and DEVICE_NOT_FOUND, and I have applied the latest only to operations that may expect a device (and also 422 applies to them)

I agree that it is clearer if it is visible which examples are belonging to which operation. But in https://github.com/camaraproject/Commonalities/blob/a4f409d3c9e278bd30afc838cee8a261ff450939/artifacts/CAMARA_common.yaml#L209 Generic404 is defined with both NOT_FOUND and DEVICE_NOT_FOUND and even a placeholder for specific codes. But ok, I will apply the approach from your #299.

jlurien commented:
> > * org.camaraproject.qod.v0.qos-status-changed -> org.camaraproject.quality-on-demand.vwip.qos-status-changed
> 
> That outside of the scope of this PR, at least we should open an issue for that to make the change more transparent. Is there a guideline regarding the versioning of events which is different from the API versioning? Otherwise it would be in the release v0.11rc1, but I'm not sure if this is correct.

Goog question. I though it was intended to match the major version of the API, but I guess the major version of the API could change without evolving the event version (but not the opposite).

hdamker commented:
> > > * org.camaraproject.qod.v0.qos-status-changed -> org.camaraproject.quality-on-demand.vwip.qos-status-changed
> > 
> > 
> > That outside of the scope of this PR, at least we should open an issue for that to make the change more transparent. Is there a guideline regarding the versioning of events which is different from the API versioning? Otherwise it would be in the release v0.11rc1, but I'm not sure if this is correct.
> 
> Good question. I though it was intended to match the major version of the API, but I guess the major version of the API could change without evolving the event version (but not the opposite).

Created #332 for the larger topic that we missed the changes within the Event subscription model, which impacts also our implicit subscriptions. Changing the event name is the smaller problem here (btw: I would stay here with v0, not using "wip" in between - the risk is too high to oversee this in the release PR. The version of the event will get only "visible", when there is a released API Version).

hdamker commented:
> > * DurationOutOfRangeForQoSProfile should not be part of Generic400 as that one should only contain the common ones. A separare 400 response would be needed for the operations that receive a duration
> > * In [New API QOD Provisioning #299](https://github.com/camaraproject/QualityOnDemand/pull/299) I have considered a Generic404 with just NOT_FOUND and GenericDevice404 with both NOT_FOUND and DEVICE_NOT_FOUND, and I have applied the latest only to operations that may expect a device (and also 422 applies to them)
> 
> I agree that it is clearer if it is visible which examples are belonging to which operation. But in https://github.com/camaraproject/Commonalities/blob/a4f409d3c9e278bd30afc838cee8a261ff450939/artifacts/CAMARA_common.yaml#L209 Generic404 is defined with both NOT_FOUND and DEVICE_NOT_FOUND and even a placeholder for specific codes. But ok, I will apply the approach from your #299.

@jlurien Done in https://github.com/camaraproject/QualityOnDemand/pull/326/commits/55c609a4a07735654f6c6e985ccac92156fc584d. Please check.

hdamker commented:
@jlurien @maxl2287 @eric-murray @RandyLevensalor The PR should be ready to be merged. Would be also good to get it merged soon, so that we can base the other PRs on it. Can you do a timely final review?

jlurien reviewed (dismissed):
LGTM, I assume #335 as a separate one but it is quite independent so merging order should not be relevant

RandyLevensalor commented on code/API_definitions/quality-on-demand.yaml:
```diff
@@ -729,7 +681,8 @@ components:
         * `networkAccessIdentifier`
 
-        NOTE: the MNO might support only a subset of these options. The API invoker can provide multiple identifiers to be compatible across different MNOs. In this case the identifiers MUST belong to the same device
+        NOTE1: the MNO might support only a subset of these options. The API invoker can provide multiple identifiers to be compatible across different MNOs. In this case the identifiers MUST belong to the same device
```
Can we remove MNO, since it is mobile specific?

hdamker replied:
Sure, it's anyway part of the alignment with Commonalities 0.4.0-rc.2 alignment. Done in https://github.com/camaraproject/QualityOnDemand/pull/326/commits/70ea70ee9386eebd4b7e102887021739cedb1385

RandyLevensalor reviewed (approved):
LGTM

hdamker commented:
Given the two approvals from @RandyLevensalor and @jlurien (before the last trivial change) I will merge this one now, so that we can continue to align the other open PRs (if needed). 

Please open issues if there other alignment points with Commonalities are open.
